doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,170 @@
+and [#10173](https://github.com/MaterializeInc/materialize/pull/10173) are previous pull requests that try and implement
+this query blocking.
+
+This approach guarantees that when a read completes, the global timestamp is equal to or larger than the timestamp of
+that read.
+
+## Alternatives
+
+- We don't hold back compaction and just block all queries that require a timestamp larger than the global timestamp.
+- If we didn't hold back compaction, then we could always advance the global timestamp to the most recent operation.
+  Then any operation on objects that haven't caught up to the global timestamp must be blocked until the object catches
+  up to the global timestamp.
+- Instead of using the wall time on each node as a proxy for the global timestamp, the Coordinator could broadcast the
+  global timestamp every time it is updated.
+- Instead of using the wall time on each node as a proxy for the global timestamp, the Coordinator can include the
+  current global timestamp in every message sent to dataflow nodes. The nodes would then prevent compaction up to the
+  most recently seen global timestamp. After the node learns of a new timestamp they can allow compaction of every
+  object up to the new timestamp. However, if a node doesn't receive any messages from the Coordinator then memory usage
+  will balloon up until it hears from the Coordinator again.
+- Instead of using group commit/write coalesce, the coordinator could buffer all writes to user tables in a durable
+  write ahead log. All writes to user tables that occur in the same millisecond can be assigned the same timestamp, as
+  long as there are no reads in-between them. Then at the end of the millisecond, or after X milliseconds, all writes
+  are sent to STORAGE in a batch. If Materialize crashes/restarts before sending the writes to STORAGE, then it can just
+  read the unsent writes from the durable logs after starting back up. This prevents us from having to increment the
+  global timestamp on every write to a user table. However, this does not solve the problem if there are reads
+  in-between the writes or if we read from a view that's ahead of the wall clock.
+- Hybrid Logical Timestamps are timestamps where the X most significant bits are used for the epoch and the Y least
+  significant bits are used for a counter. Every clock tick the X bits are increased by 1 and the Y bytes are all
+  reset to 0. Every write increments the Y bits by 1. This would allow us to have 2^Y writes per clock tick per object,
+  without blocking any queries. Any write in excess of 2^Y per clock tick per object must be blocked until the next
+  clock tick. This way no object's `upper` would have to advance past the current global timestamp if the global
+  timestamp tracked the wall clock. CockroachDB uses similar timestamps
+  here: https://github.com/cockroachdb/cockroach/blob/711ccb5b8fba4e6a826ed6ba46c0fc346233a0f7/pkg/sql/sem/builtins/builtins.go#L8548-L8560
+
+## Open Questions
```

Comment:
I haven't though about this much but here are some of my initial thoughts:
- I don't have a strong opinion on whether or not strict serializability should be the default and users need to opt out, or the current behavior should be the default and users need to opt in. I think we should see what the performance implications of strict serializability are and use that to inform that decision.
- I believe it should be possible to mix and match consistency levels. 
  - Non-strict serializable reads can just pick any timestamp that satisfies the query ( `since <= ts < upper`), without caring about the global timestamp. 
  - Non-strict serializable writes can be assigned some timestamp larger than the global timestamp without advancing the global timestamp.

Reply:
I think its important to consider if a specific strict-serializable proposal allows strict reads/writes to affect the performance of non-struct ones, it sounds like we have considered that. I would be curious to know how this works in the write case, can you pick a time `> now()`?

Reply:
I have been thinking about this more, for reads I'm still confident that my approach above will work fine. However, I don't actually think that you can mix non-strict serializable writes with strict serializable writes. 
For example if you have a non-strict serializable write commit and we pick a time `> now()`, then we have a strict serializable read, should we guarantee the read should see the non-strict serializable write? I think so, which would mean doing something like blocking the read until the global timestamp advances to that whatever time we picked for the write.

Reply:
yeah, I think you are right, we have to block the read till it sees the write. However, non-strict affecting strict is likely less bad than strict affecting non-strict, which this proposal probably doesnt cause?

Reply:
strict writes do not affect non-strict reads and strict reads do not affect non-strict reads. However this proposal does suggest changing all writes such that they need to block until the next group commit (though this could be a per deployment command line arg).